### ISSUE TITLE
[Question]: Creating parallel tests from a collection while reusing browser resources

### ISSUE THREAD

**Issue by ChristianMay21** [microsoft/playwright-dotnet#1721]
### Your question

Hi! Thanks for all the great work on Playwright, it's a really great tool.

Quick question: I would like to use Playwright to create a scheduled task that loads every page on a site, just to check that none of them are broken.

The way I am trying to achieve that is by using NUnit's TestCaseSource attribute to create a test from every item in a collection of strings, where each string is a URL on the page.

Run serially, one URL at a time, this works great. However, I would really love to be able to 1) run them in parallel, and 2) reuse browsers, at least to some extent. Here's what I have right now:

`using Microsoft.Playwright.NUnit;
using NUnit.Framework;
using System.Threading.Tasks;
using Microsoft.Playwright;
using System.Collections.Concurrent;
using System.Linq;

//The purpose of this test is to test every page in the sitemap to see if any pages are broken
namespace PlaywrightWebTesting
{
	
	[TestFixture, Parallelizable(ParallelScope.Children)]
	public class MyTest : BrowserTest
	{
		public static readonly ConcurrentQueue<string> sitemapPages;
		static MyTest()
		{
			sitemapPages = new ConcurrentQueue<string>();
			foreach(var i in Enumerable.Range(0, 40))
			{
				sitemapPages.Enqueue("https://www.google.com");
			}
		}

		//browser should be passed to method since browsers are expensive to create - new contexts and pages are cheap to create, so they will be created in the method
		[Test, TestCaseSource(nameof(sitemapPages))]
		public async Task TestThatPageLoads(string pageURL)
		{
			var context = await Browser.NewContextAsync();
	        var page = await context.NewPageAsync();

			//Go to URL to test
	        await page.GotoAsync(pageURL, new PageGotoOptions {Timeout = 0 });

			// Close page
			await page.CloseAsync();
			await context.CloseAsync();
		}
	}
}`

Note that, for minimum reproduction purposes, instead of a sitemap I'm just using a ConcurrentQueue with the site google.com 40 times. 

When I run this test, I successfully get 40 tests trying to ping google, but about half the time I'll get one test which gets the following error:

![image](https://user-images.githubusercontent.com/45889402/131571144-852108b0-d48c-418c-aa0d-2531e281a79c.png)
`    30)  TestThatPageLoads("https://www.google.com")
      Duration: 12.7 sec

      Message: 
        TearDown : System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
        TearDown : System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
      Stack Trace: 
        --TearDown
        Enumerator.MoveNextRare()
        BrowserTest.BrowserTearDown()
        GenericAdapter`1.GetResult()
        AsyncToSyncAdapter.Await(Func`1 invoke)
        SetUpTearDownItem.RunSetUpOrTearDownMethod(TestExecutionContext context, MethodInfo method)
        SetUpTearDownItem.RunTearDown(TestExecutionContext context)
        --TearDown
        Enumerator.MoveNext()
        WorkerAwareTest.WorkerTeardown()
        GenericAdapter`1.BlockUntilCompleted()
        AsyncToSyncAdapter.Await(Func`1 invoke)
        SetUpTearDownItem.RunSetUpOrTearDownMethod(TestExecutionContext context, MethodInfo method)
        SetUpTearDownItem.RunTearDown(TestExecutionContext context)`

This error, presumably, crashes the browser shared by the tests, because then a bunch of other tests will fail because the "browser has unexpectedly closed".

I can make this software pattern (using TestCaseSource) work if I change the class to an instance of PlaywrightTest, rather than BrowserTest, and spin up a new browser for each test, but that's an expensive operation that I'd like to avoid running on every test if at all possible.

I have made a minimum reproduction available at https://github.com/ChristianMay21/Playwright-Parallel-Repro/blob/main/PageTests.cs

**Comment by avodovnik** [comment]
Our harness libraries are created with parallelizing on the class level in mind, so this won't work.

In your case, I *think* something like this might do the trick:

```csharp
using Microsoft.Playwright;
using NUnit.Framework;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace TestProject1
{
    [TestFixture, Parallelizable(ParallelScope.Children)]
    public class Tests
    {
        public static readonly List<string> sitemapPages = new();
        static Tests()
        {
            for(int i = 0; i < 40; i++)
                sitemapPages.Add("https://www.google.com");
        }

        private readonly Stack<(IPage page, int id)> pagePool = new();
        private const int MaxPageCount = 10;
        private readonly Semaphore pageSemaphore = new(0, MaxPageCount);

        private IPlaywright playwright;
        private IBrowser browser;
        private bool isSetup = false;
        [OneTimeSetUp]
        public async Task Setup()
        {
            playwright = await Playwright.CreateAsync();
            browser = await playwright.Chromium.LaunchAsync();

            lock (pagePool)
            {
                if (isSetup) return;
                isSetup = true;
            }

            int i = 0;
            while(i++ < MaxPageCount)
            {
                pagePool.Push((await browser.NewPageAsync(), i));
            }

            pageSemaphore.Release(MaxPageCount);
        }

        [Test, TestCaseSource(nameof(sitemapPages))]
        public async Task Test(string pageUrl)
        {
            pageSemaphore.WaitOne();

            var tmp = pagePool.Pop();
            var page = tmp.page;
            System.Console.WriteLine($"Working on page {tmp.id}");
            var req = await page.GotoAsync(pageUrl);
            Assert.IsTrue(req.Status == 200);
            pagePool.Push(tmp);

            pageSemaphore.Release();
        }

        [OneTimeTearDown]
        public async Task TearDown()
        {
            if(browser != null)
                await browser.DisposeAsync();
            playwright?.Dispose();
        }
    }
}
```

**Comment by ChristianMay21** [comment]
Oh awesome - I'll take this for a spin when I've got a time at the end of my current sprint and report back here - thanks so much for the help!

**Comment by mxschmitt** [comment]
Closing as part of the triage process since it seemed stale. Please create a new issue with a detailed reproducible or feature request if you still face issues.
